File: source/java-sdk/src/main/java/com/aerospike/examples/gaming/PlayerMatching.java
package com.aerospike.examples.gaming;

import java.time.Duration;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Optional;
import java.util.concurrent.ThreadLocalRandom;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;

import com.aerospike.client.sdk.AerospikeException;
import com.aerospike.client.sdk.ChainableOperationBuilder;
import com.aerospike.client.sdk.DataSet;
import com.aerospike.client.sdk.Record;
import com.aerospike.client.sdk.RecordResult;
import com.aerospike.client.sdk.Session;
import com.aerospike.client.sdk.TypedDataSet;
import com.aerospike.client.sdk.TypedKey;
import com.aerospike.client.sdk.TypedKeyList;
import com.aerospike.examples.Async;
import com.aerospike.examples.UseCase;
import com.aerospike.examples.gaming.model.Player;

/**
 * SDK port of the legacy {@code PlayerMatching} (see ../../java). Reuses {@link Leaderboard}'s
 * scoreboard update/query logic for scoring, and layers matchmaking on top via filter expressions
 * ({@code where(Exp)}) instead of the legacy client's {@code Policy.filterExp}.
 * <p/>
 * The legacy version's {@code setPlayerOnline} builds its conditional-write filter against a
 * bin called {@code "isOnline"}, which is never written anywhere (the actual online-state bin is
 * {@code "online"}) - so that filter can never match. This port fixes it to check {@code "online"},
 * matching the method's own documented intent (refuse to re-mark an already-online player online).
 */
public class PlayerMatching implements UseCase {

    private static final int NUM_PLAYERS = 10_000;
    private static final int RUNTIME_SECS = 10;
    private static final long SHIELD_DURATION_MS = TimeUnit.SECONDS.toMillis(5);

    private static final String[] FIRST_NAMES = {"Estefana", "Arnoldo", "Jacquelyne", "Harris", "Ardelia"};
    private static final String[] LAST_NAMES = {"Ruecker", "MacGyver", "Willms", "Jones", "Renner"};

    private final Leaderboard leaderboard = new Leaderboard();

    @Override
    public String getName() {
        return "Player matching";
    }

    @Override
    public String getReference() {
        return "https://github.com/aerospike-examples/use-case-cookbook/blob/main/UseCases/player-matching.md";
    }

    @Override
    public String getDescription() {
        return "Find players that match certain criteria at scale. Players play the game concurrently, but in order "
                + "to attack they need to match against another player. The player cannot be online in this example, "
                + "as they may be playing the game. In order to match, we want another player who:\n"
                + "1. Is not online\n"
                + "2. Who does not have a shield\n"
                + "3. Who is not currently being attacked\n"
                + "4. Whose score is >= 400\n"
                + "5. Whose score is similar to this player\n"
                + "A shield is given to a player when they have been attacked and defeated. These would normally be a decent length "
                + "to stop them being attacked too often, but in this example will only be for 5s. Attackers defeat "
                + "their opponents 80% of the time. If a player with a shield attacks during a shield, the shield is removed.";
    }

    private final TypedDataSet<Player> players =
            TypedDataSet.of(System.getProperty("demo.namespace", "test"), "uccb_player", Player.class);
    private final DataSet scoreboard = DataSet.of(System.getProperty("demo.namespace", "test"), "scoreboard");

    private Player randomPlayer(int id) {
        String first = FIRST_NAMES[ThreadLocalRandom.current().nextInt(FIRST_NAMES.length)];
        String last = LAST_NAMES[ThreadLocalRandom.current().nextInt(LAST_NAMES.length)];
        int score = ThreadLocalRandom.current().nextInt(0, 6201);
        return new Player(id, first.toLowerCase() + last.toLowerCase(), first, last,
                first.toLowerCase() + "." + last.toLowerCase() + "@example.com",
                0, false, "", score);
    }

    @Override
    public void setup(Session session) throws Exception {
        session.truncate(players);
        session.truncate(scoreboard);

        System.out.printf("Generating %,d Players%n", NUM_PLAYERS);
        for (int id = 1; id <= NUM_PLAYERS; id++) {
            Player player = randomPlayer(id);
            session.upsert(players).object(player).execute();
            leaderboard.updatePlayerScore(session, player.getId(), -1, player.getScore());
        }
    }

    @Override
    public void run(Session session) throws Exception {
        testEligibility(session);

        System.out.printf("%nLet's play some games!%n");
        Player player1 = setPlayerOnline(session, 1, true);
        System.out.println("Leader board before the games start...");
        leaderboard.showPlayersAroundPlayer(session, player1.getId(), player1.getScore());

        AtomicInteger counter = new AtomicInteger();
        Async.runFor(Duration.ofSeconds(RUNTIME_SECS), (async) -> {
            async.periodic(Duration.ofMillis(100), () -> {
                findPlayerToAttack(session, player1).ifPresent(defender -> {
                    playGame(session, player1, defender, true);
                    counter.incrementAndGet();
                });
            });

            async.periodic(Duration.ofMillis(5), 20, () -> {
                int playerId = async.rand().nextInt(NUM_PLAYERS - 1) + 2;
                Player player = setPlayerOnline(session, playerId, true);
                if (player != null) {
                    try {
                        findPlayerToAttack(session, player).ifPresent(defender -> {
                            playGame(session, player, defender, false);
                            counter.incrementAndGet();
                        });
                    }
                    finally {
                        setPlayerOnline(session, playerId, false);
                    }
                }
            });
        });
        System.out.println("Leader board after the games end...");
        leaderboard.showPlayersAroundPlayer(session, player1.getId(), player1.getScore());

        System.out.printf("Total battles played (all threads): %,d%n", counter.get());
        setPlayerOnline(session, player1.getId(), false);
    }

    /**
     * Finds a player to attack from a list of candidate keys: batch-reads them filtered down to
     * eligible players, then tries each (randomly) with a filtered conditional write that claims
     * them (sets {@code beingAttackedBy}) atomically, retrying with another candidate if the claim
     * fails because the candidate stopped being eligible in the meantime.
     */
    public Optional<Player> findPlayerToAttack(Session session, int attackerId, TypedKeyList<Player> possibilities) {
        String filter = getPlayerFilter();
        List<Record> validRecords = new ArrayList<>();
        try (var stream = session.query(possibilities).where(filter).execute()) {
            stream.forEach(result -> {
                if (result.isOk()) {
                    validRecords.add(result.recordOrThrow());
                }
            });
        }

        while (!validRecords.isEmpty()) {
            int recordNum = ThreadLocalRandom.current().nextInt(validRecords.size());
            int id = validRecords.get(recordNum).getInt("id");
            // Note: "beingAttackedBy" is set above but deliberately not also read back below -
            // requesting both a write and a read of the same bin in one operation returns a
            // multi-result wrapper for that bin instead of a plain value, so we just set it
            // directly on the Player below from the value we already know we wrote.
            //
            // Decoded manually (not via the mapper) because this is an upsert-with-read-back, not
            // a query: AeroRecordMapper.fromMap(Map, Key, int) - the 3-arg overload RecordStream
            // .getFirst(RecordMapper<T>) calls here - unconditionally throws
            // UnsupportedOperationException ("JOM requires a RecordReadContext for dependency
            // resolution"), even for a dependency-free class like Player. The mapper only supports
            // decoding through the SDK's typed query path (TypedKeyQueryBuilder/TypedRecordStream,
            // which does pass that context), not this kind of write-then-read-selected-bins result.
            Optional<RecordResult> result = session.upsert(getPlayerKey(id))
                    .where(filter)
                    .bin("beingAttackedBy").setTo("Player " + id)
                    .bin("id").get()
                    .bin("userName").get()
                    .bin("firstName").get()
                    .bin("lastName").get()
                    .bin("email").get()
                    .bin("shieldExpiry").get()
                    .bin("online").get()
                    .bin("score").get()
                    .execute().getFirst();

            if (result.isEmpty() || !result.get().isOk()) {
                validRecords.remove(recordNum);
            }
            else {
                Player player = recordToPlayer(result.get().recordOrThrow());
                player.setBeingAttackedBy("Player " + id);
                return Optional.of(player);
            }
        }
        return Optional.empty();
    }

    /** Given an attacker, finds a player of similar strength who is available to attack. */
    public Optional<Player> findPlayerToAttack(Session session, Player attacker) {
        List<Player> similarScores = leaderboard.getScoresAroundPlayer(session, attacker.getId(), attacker.getScore(), 20);
        TypedKeyList<Player> keys = new TypedKeyList<>();
        similarScores.stream()
                .filter(player -> player.getId() != attacker.getId())
                .forEach(player -> keys.add(getPlayerKey(player.getId())));
        return findPlayerToAttack(session, attacker.getId(), keys);
    }

    public static int calculateNewEloRating(int playerRating, int opponentRating, int score, int kFactor) {
        double expectedScore = 1.0 / (1.0 + Math.pow(10.0, (opponentRating - playerRating) / 400.0));
        double newRating = playerRating + kFactor * (score - expectedScore);
        return (int) Math.round(newRating);
    }

    /** Plays a game between an attacker and a defender, adjusting scores, shield, and leaderboard. */
    public void playGame(Session session, Player attacker, Player defender, boolean showBattle) {
        if (attacker == null || defender == null) {
            return;
        }

        if (showBattle) {
            System.out.printf("%s is attacking %s... ", attacker.getUserName(), defender.getUserName());
        }
        try {
            Thread.sleep(5);
        }
        catch (InterruptedException ignored) {
        }
        int originalAttackerScore = attacker.getScore();
        int originalDefenderScore = defender.getScore();

        int probabilityOfWinning = 60 + (attacker.getScore() - defender.getScore()) / 5;
        if (ThreadLocalRandom.current().nextInt(101) >= probabilityOfWinning) {
            attacker.setScore(calculateNewEloRating(attacker.getScore(), defender.getScore(), 1, 20));
            defender.setScore(calculateNewEloRating(defender.getScore(), attacker.getScore(), 0, 20));
            defender.setShieldExpiry(new Date().getTime() + SHIELD_DURATION_MS);
            if (showBattle) {
                System.out.printf("VICTORIOUS! (attacker: %,d -> %,d, defender: %,d->%,d)%n",
                        originalAttackerScore, attacker.getScore(),
                        originalDefenderScore, defender.getScore());
            }
        }
        else {
            attacker.setScore(calculateNewEloRating(attacker.getScore(), defender.getScore(), 0, 20));
            defender.setScore(calculateNewEloRating(defender.getScore(), attacker.getScore(), 1, 20));
            if (showBattle) {
                System.out.printf("REPELLED! (attacker: %,d -> %,d, defender: %,d->%,d)%n",
                        originalAttackerScore, attacker.getScore(),
                        originalDefenderScore, defender.getScore());
            }
        }
        attacker.setShieldExpiry(0);
        defender.setBeingAttackedBy(null);

        session.doInTransaction(tx -> {
            leaderboard.updatePlayerScore(tx, attacker.getId(), originalAttackerScore, attacker.getScore());
            leaderboard.updatePlayerScore(tx, defender.getId(), originalDefenderScore, defender.getScore());

            tx.upsert(getPlayerKey(attacker.getId()))
                    .bin("score").setTo(attacker.getScore())
                    .bin("shieldExpiry").setTo(attacker.getShieldExpiry())
                    .execute();

            tx.upsert(getPlayerKey(defender.getId()))
                    .bin("score").setTo(defender.getScore())
                    .bin("shieldExpiry").setTo(defender.getShieldExpiry())
                    .bin("beingAttackedBy").setTo("")
                    .execute();
        });
    }

    /**
     * Every {@code Player} always has {@code beingAttackedBy} set at creation, so (unlike the
     * legacy filter) this doesn't need a missing-bin fallback clause.
     */
    private String getPlayerFilter() {
        return String.format(
                "$.online == false and $.shieldExpiry < %d and $.beingAttackedBy == '' and $.score > 400",
                new Date().getTime());
    }

    public TypedKey<Player> getPlayerKey(int id) {
        return players.id(id);
    }

    /**
     * Manually decodes a {@code Player} from a {@code Record} of the selected bins. Needed
     * because {@code AeroRecordMapper}'s 3-arg {@code fromMap(Map, Key, int)} - the overload
     * {@code RecordStream.getFirst(RecordMapper<T>)} calls - unconditionally throws
     * {@code UnsupportedOperationException("JOM requires a RecordReadContext for dependency
     * resolution")}, even for a dependency-free class like {@code Player}. The mapper only
     * decodes through the SDK's typed query path (which does supply that context), not an
     * upsert-with-read-back result like {@link #findPlayerToAttack} and {@link #setPlayerOnline}
     * produce.
     */
    private Player recordToPlayer(Record rec) {
        return new Player(rec.getInt("id"), rec.getString("userName"), rec.getString("firstName"),
                rec.getString("lastName"), rec.getString("email"), rec.getLong("shieldExpiry"),
                rec.getBoolean("online"), rec.getString("beingAttackedBy"), rec.getInt("score"));
    }

    /**
     * Sets a player online and returns their details, or {@code null} if {@code isOnline} is true
     * and the player was already online.
     */
    public Player setPlayerOnline(Session session, int playerId, boolean isOnline) {
        // A where clause of "" applies no filter, so the isOnline/not-isOnline cases don't need
        // separate builder-reassignment branches.
        // "online" is set above but deliberately not also read back below - see the equivalent
        // note in findPlayerToAttack. Decoded manually rather than via the mapper - see
        // recordToPlayer's javadoc for why.
        Optional<RecordResult> result = session.upsert(getPlayerKey(playerId))
                .where(isOnline ? "$.online == false" : "")
                .bin("online").setTo(isOnline)
                .bin("id").get()
                .bin("userName").get()
                .bin("firstName").get()
                .bin("lastName").get()
                .bin("email").get()
                .bin("shieldExpiry").get()
                .bin("beingAttackedBy").get()
                .bin("score").get()
                .execute().getFirst();
        if (result.isEmpty() || !result.get().isOk()) {
            return null;
        }
        Player player = recordToPlayer(result.get().recordOrThrow());
        player.setOnline(isOnline);
        return player;
    }

    /** Resets a player to offline, no shield, not being attacked. Sets the score too if >= 0. */
    public void resetPlayerTo(Session session, int playerId, int score) {
        ChainableOperationBuilder op = session.upsert(getPlayerKey(playerId))
                .bin("online").setTo(false)
                .bin("shieldExpiry").setTo(0L)
                .bin("beingAttackedBy").setTo("");
        if (score >= 0) {
            op = op.bin("score").setTo(score);
        }
        op.execute();
    }

    /**
     * Determines if the player can currently be attacked. Not for real game play - the result
     * could be stale by the time it's returned.
     */
    public boolean canAttackPlayerTest(Session session, int playerId) {
        try {
            Optional<RecordResult> result = session.query(getPlayerKey(playerId))
                    .where(getPlayerFilter())
                    .failOnFilteredOut()
                    .execute().getFirst();
            if (result.isEmpty()) {
                System.out.println("*** Key " + playerId + " does not exist!");
                return false;
            }
            result.get().recordOrThrow();
            return true;
        }
        catch (AerospikeException.FilteredException fe) {
            return false;
        }
    }

    public int testEligibility(Session session) {
        int playerId = 1;
        Record playerRecord = session.query(getPlayerKey(playerId)).execute().getFirstRecord();
        if (playerRecord == null) {
            throw new IllegalStateException(
                    "No player with id " + playerId + " found - run setup() before run()/testEligibility().");
        }
        int originalScore = playerRecord.getInt("score");
        System.out.printf("%nTesting eligibility for player %d%n", playerId);
        resetPlayerTo(session, playerId, 590);
        System.out.printf("- Checking player can validly be attacked: %s%n",
                canAttackPlayerTest(session, playerId) ? "PASSED" : "FAILED");

        setPlayerOnline(session, playerId, true);
        System.out.printf("- Checking player cannot be attacked when online: %s%n",
                !canAttackPlayerTest(session, playerId) ? "PASSED" : "FAILED");

        resetPlayerTo(session, playerId, 50);
        System.out.printf("- Checking player cannot be attacked with a low score: %s%n",
                !canAttackPlayerTest(session, playerId) ? "PASSED" : "FAILED");

        resetPlayerTo(session, playerId, 500);
        leaderboard.updatePlayerScore(session, playerId, originalScore, 1234);
        return 1234;
    }
}

File: source/python-sdk/usecasecookbook/gaming/player_matching.py
"""Port of ../../java-sdk's PlayerMatching (itself a port of ../../java's, see that
class's javadoc). Reuses :class:`~usecasecookbook.gaming.leaderboard.Leaderboard`'s
scoreboard update/query logic for scoring, and layers matchmaking on top via AEL
``where()`` filters for conditional (compare-and-swap style) writes.

Simpler than ../../java-sdk here: this SDK has no object mapper (every model is decoded
by hand via ``Player.from_bins``), and reading back a bin in the same call it was just
written returns a plain scalar, not a wrapper - so this port just reads back every bin it
needs, written or not, in one round trip.
"""

import random
from datetime import datetime, timezone

from aerospike_async import Key
from aerospike_async.exceptions import ResultCode
from aerospike_sdk.exceptions import AerospikeError
from aerospike_sdk.sync import Session

from usecasecookbook.async_util import Async
from usecasecookbook.gaming.leaderboard import PLAYERS, SCOREBOARD, Leaderboard
from usecasecookbook.gaming.model import Player
from usecasecookbook.txn import run_in_transaction
from usecasecookbook.use_case import UseCase

NUM_PLAYERS = 10_000
RUNTIME_SECS = 10
SHIELD_DURATION_MS = 5_000

FIRST_NAMES = ["Estefana", "Arnoldo", "Jacquelyne", "Harris", "Ardelia"]
LAST_NAMES = ["Ruecker", "MacGyver", "Willms", "Jones", "Renner"]


def _random_player(id: int) -> Player:
    first = random.choice(FIRST_NAMES)
    last = random.choice(LAST_NAMES)
    score = random.randint(0, 6200)
    return Player(
        id=id, user_name=f"{first.lower()}{last.lower()}", first_name=first, last_name=last,
        email=f"{first.lower()}.{last.lower()}@example.com",
        shield_expiry=0, online=False, being_attacked_by="", score=score,
    )


def calculate_new_elo_rating(player_rating: int, opponent_rating: int, score: int, k_factor: int) -> int:
    """Calculates the new Elo rating for a player after a game.

    Args:
        player_rating: The current rating of the player (e.g., 1600).
        opponent_rating: The rating of the opponent (e.g., 1800).
        score: The actual score of the game for the player: 1 = win, 0 = draw or loss.
        k_factor: The K-factor used to determine rating volatility (e.g., 20).

    Returns:
        The new rating of the player after the game, rounded to the nearest integer.
    """
    expected_score = 1.0 / (1.0 + pow(10.0, (opponent_rating - player_rating) / 400.0))
    new_rating = player_rating + k_factor * (score - expected_score)
    return round(new_rating)


class PlayerMatching(UseCase):
    def __init__(self) -> None:
        self.leaderboard = Leaderboard()

    def get_name(self) -> str:
        return "Player matching"

    def get_reference(self) -> str:
        return "https://github.com/aerospike-examples/use-case-cookbook/blob/main/UseCases/player-matching.md"

    def get_description(self) -> str:
        return (
            "Find players that match certain criteria at scale. Players play the game concurrently, but in order "
            "to attack they need to match against another player. The player cannot be online in this example, "
            "as they may be playing the game. In order to match, we want another player who:\n"
            "1. Is not online\n"
            "2. Who does not have a shield\n"
            "3. Who is not currently being attacked\n"
            "4. Whose score is >= 400\n"
            "5. Whose score is similar to this player\n"
            "A shield is given to a player when they have been attacked and defeated. These would normally be a decent length "
            "to stop them being attacked too often, but in this example will only be for 5s. Attackers defeat "
            "their opponents 80% of the time. If a player with a shield attacks during a shield, the shield is removed."
        )

    def setup(self, session: Session) -> None:
        session.truncate(PLAYERS)
        session.truncate(SCOREBOARD)

        print(f"Generating {NUM_PLAYERS:,} Players")
        for id in range(1, NUM_PLAYERS + 1):
            player = _random_player(id)
            session.upsert(PLAYERS.id(player.id)).put(player.to_bins()).execute()
            self.leaderboard.update_player_score(session, player.id, -1, player.score)

    def run(self, session: Session) -> None:
        self.test_eligibility(session)

        print("\nLet's play some games!")
        player1 = self.set_player_online(session, 1, True)
        print("Leader board before the games start...")
        self.leaderboard.show_players_around_player(session, player1.id, player1.score)

        counter = [0]

        def run_simulation(async_runner: Async) -> None:
            def attacker_loop() -> None:
                defender = self.find_player_to_attack(session, player1)
                if defender is not None:
                    self.play_game(session, player1, defender, True)
                    counter[0] += 1
            async_runner.periodic(0.1, attacker_loop)

            def random_player_loop() -> None:
                player_id = random.randint(2, NUM_PLAYERS)
                player = self.set_player_online(session, player_id, True)
                if player is not None:
                    try:
                        defender = self.find_player_to_attack(session, player)
                        if defender is not None:
                            self.play_game(session, player, defender, False)
                            counter[0] += 1
                    finally:
                        self.set_player_online(session, player_id, False)
            async_runner.periodic(0.005, random_player_loop, 20)

        Async.run_for(RUNTIME_SECS, run_simulation)

        print("Leader board after the games end...")
        self.leaderboard.show_players_around_player(session, player1.id, player1.score)

        print(f"Total battles played (all threads): {counter[0]:,}")
        self.set_player_online(session, player1.id, False)

    def find_player_to_attack_by_keys(
        self, session: Session, attacker_id: int, possibilities: list[Key],
    ) -> Player | None:
        """Finds a player to attack from a list of candidate keys: batch-reads them
        filtered down to eligible players, then tries each (randomly) with a filtered
        conditional write that claims them (sets ``beingAttackedBy``) atomically,
        retrying with another candidate if the claim fails because the candidate
        stopped being eligible in the meantime.
        """
        if not possibilities:
            return None
        filter_ael = self._player_filter()
        valid_records = []
        stream = session.query(possibilities).where(filter_ael).execute()
        for row in stream:
            if row.is_ok and row.record is not None:
                valid_records.append(row.record.bins)
        stream.close()

        while valid_records:
            record_num = random.randrange(len(valid_records))
            id = valid_records[record_num]["id"]

            row = (
                session.upsert(self.get_player_key(id))
                .where(filter_ael)
                .bin("beingAttackedBy").set_to(f"Player {id}")
                .bin("id").get()
                .bin("userName").get()
                .bin("firstName").get()
                .bin("lastName").get()
                .bin("email").get()
                .bin("shieldExpiry").get()
                .bin("online").get()
                .bin("beingAttackedBy").get()
                .bin("score").get()
                .execute().first()
            )

            if row is None or not row.is_ok or row.record is None:
                valid_records.pop(record_num)
            else:
                return Player.from_bins(row.record.bins)
        return None

    def find_player_to_attack(self, session: Session, attacker: Player) -> Player | None:
        """Given an attacker, finds a player of similar strength who is available to
        attack. The player will already have been locked for attacking.
        """
        similar_scores = self.leaderboard.get_scores_around_player(session, attacker.id, attacker.score, 20)
        keys = [self.get_player_key(e.id) for e in similar_scores if e.id != attacker.id]
        return self.find_player_to_attack_by_keys(session, attacker.id, keys)

    def play_game(self, session: Session, attacker: Player, defender: Player, show_battle: bool) -> None:
        """Plays a game between an attacker and a defender, adjusting scores, shield,
        and leaderboard.
        """
        if attacker is None or defender is None:
            return

        if show_battle:
            print(f"{attacker.user_name} is attacking {defender.user_name}... ", end="")

        original_attacker_score = attacker.score
        original_defender_score = defender.score

        probability_of_winning = 60 + (attacker.score - defender.score) // 5
        if random.randint(0, 100) >= probability_of_winning:
            attacker.score = calculate_new_elo_rating(attacker.score, defender.score, 1, 20)
            defender.score = calculate_new_elo_rating(defender.score, attacker.score, 0, 20)
            defender.shield_expiry = int(datetime.now(timezone.utc).timestamp() * 1000) + SHIELD_DURATION_MS
            if show_battle:
                print(
                    f"VICTORIOUS! (attacker: {original_attacker_score:,} -> {attacker.score:,}, "
                    f"defender: {original_defender_score:,}->{defender.score:,})"
                )
        else:
            attacker.score = calculate_new_elo_rating(attacker.score, defender.score, 0, 20)
            defender.score = calculate_new_elo_rating(defender.score, attacker.score, 1, 20)
            if show_battle:
                print(
                    f"REPELLED! (attacker: {original_attacker_score:,} -> {attacker.score:,}, "
                    f"defender: {original_defender_score:,}->{defender.score:,})"
                )
        attacker.shield_expiry = 0
        defender.being_attacked_by = None

        def op(tx: Session) -> None:
            self.leaderboard.update_player_score(tx, attacker.id, original_attacker_score, attacker.score)
            self.leaderboard.update_player_score(tx, defender.id, original_defender_score, defender.score)

            (
                tx.upsert(self.get_player_key(attacker.id))
                .bin("score").set_to(attacker.score)
                .bin("shieldExpiry").set_to(attacker.shield_expiry)
                .execute()
            )
            (
                tx.upsert(self.get_player_key(defender.id))
                .bin("score").set_to(defender.score)
                .bin("shieldExpiry").set_to(defender.shield_expiry)
                .bin("beingAttackedBy").set_to("")
                .execute()
            )

        run_in_transaction(session, op)

    def _player_filter(self) -> str:
        """Every ``Player`` always has ``beingAttackedBy`` set at creation, so (unlike
        the legacy filter) this doesn't need a missing-bin fallback clause.
        """
        now_ms = int(datetime.now(timezone.utc).timestamp() * 1000)
        return f"$.online == false and $.shieldExpiry < {now_ms} and $.beingAttackedBy == '' and $.score > 400"

    def get_player_key(self, id: int) -> Key:
        return PLAYERS.id(id)

    def set_player_online(self, session: Session, player_id: int, is_online: bool) -> Player | None:
        """Sets a player online and returns their details, or ``None`` if ``is_online``
        is true and the player was already online.
        """
        builder = session.upsert(self.get_player_key(player_id))
        if is_online:
            builder = builder.where("$.online == false")
        row = (
            builder
            .bin("online").set_to(is_online)
            .bin("id").get()
            .bin("userName").get()
            .bin("firstName").get()
            .bin("lastName").get()
            .bin("email").get()
            .bin("shieldExpiry").get()
            .bin("online").get()
            .bin("beingAttackedBy").get()
            .bin("score").get()
            .execute().first()
        )
        if row is None or not row.is_ok or row.record is None:
            return None
        return Player.from_bins(row.record.bins)

    def reset_player_to(self, session: Session, player_id: int, score: int) -> None:
        """Resets a player to offline, no shield, not being attacked. Sets the score
        too if ``score >= 0``.
        """
        builder = (
            session.upsert(self.get_player_key(player_id))
            .bin("online").set_to(False)
            .bin("shieldExpiry").set_to(0)
            .bin("beingAttackedBy").set_to("")
        )
        if score >= 0:
            builder = builder.bin("score").set_to(score)
        builder.execute()

    def can_attack_player_test(self, session: Session, player_id: int) -> bool:
        """Determines if the player can currently be attacked. Not for real game play -
        the result could be stale by the time it's returned.
        """
        try:
            row = (
                session.query(self.get_player_key(player_id))
                .where(self._player_filter())
                .fail_on_filtered_out()
                .execute().first()
            )
            if row is None:
                print(f"*** Key {player_id} does not exist!")
                return False
            return True
        except AerospikeError as e:
            if e.result_code == ResultCode.FILTERED_OUT:
                return False
            raise

    def test_eligibility(self, session: Session) -> int:
        player_id = 1
        row = session.query(self.get_player_key(player_id)).execute().first()
        if row is None or row.record is None:
            raise RuntimeError(
                f"No player with id {player_id} found - run setup() before run()/test_eligibility()."
            )
        original_score = row.record.bins["score"]
        print(f"\nTesting eligibility for player {player_id}")
        self.reset_player_to(session, player_id, 590)
        passed = self.can_attack_player_test(session, player_id)
        print(f"- Checking player can validly be attacked: {'PASSED' if passed else 'FAILED'}")

        self.set_player_online(session, player_id, True)
        passed = not self.can_attack_player_test(session, player_id)
        print(f"- Checking player cannot be attacked when online: {'PASSED' if passed else 'FAILED'}")

        self.reset_player_to(session, player_id, 50)
        passed = not self.can_attack_player_test(session, player_id)
        print(f"- Checking player cannot be attacked with a low score: {'PASSED' if passed else 'FAILED'}")

        self.reset_player_to(session, player_id, 500)
        self.leaderboard.update_player_score(session, player_id, original_score, 1234)
        return 1234
